Userland/Libraries/LibJS/Tests/test-common.js

```diff
@@ -29,26 +29,39 @@ class ExpectationError extends Error {
 
 // Use an IIFE to avoid polluting the global namespace as much as possible
 (() => {
-    // FIXME: This is a very naive deepEquals algorithm
     const deepEquals = (a, b) => {
-        if (Array.isArray(a)) return Array.isArray(b) && deepArrayEquals(a, b);
-        if (typeof a === "object") return typeof b === "object" && deepObjectEquals(a, b);
-        return Object.is(a, b);
+        if (a === b) return true; // Handles identical references and primitives
+        if (typeof a !== typeof b) return false;
+        if ((a !== null && b === null) || (a === null && b !== null)) return false;
+        if (Array.isArray(a) && Array.isArray(b)) {
+            return deepArrayEquals(a, b);
+        } else if (typeof a === "object" && typeof b === "object") {
+            return deepObjectEquals(a, b);
+        }
+
+        return false;
```

Comment:
There are a few problems with this change:
- `Object.is` is not the same as `===` with respect to signed zeros and `NaN`, and I believe the behavior of `Object.is` is closer to what we want
- We ensure `typeof` both values is the same, but then check both of them on line 38; we only need to check one
- If `Array.isArray(a)` is true but `Array.isArray(b)` is false, we can abort early. That was what happened before, but now we fall into the unnecessary `typeof` check, which may in fact return true and end up doing a lot of unnecesary work. 
- The `null` handling is now completely wrong; if both objects are null, we go into `deepObjectEquals`, which now no longer has a null check

There are a few things I like though:
- Moving the `Object.is` check to the front is nice, since if they are the same object then there is no need to go through all of the comparisons. 
- Now that I look at the old version, it really needs some symmetric checks, which you did to an extent. If `Array.isArray(a)` is `false`, but `Array.isArray(b)` is true, we should immediately return false. And the same thing is true for the typeof check

Reply:
Fixed 👍 

Thanks for the review!